delta-app/src/main/java/io/cdap/delta/store/RemoteHttpClient.java

```diff
@@ -0,0 +1,154 @@
+import io.cdap.delta.app.service.AssessmentService;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.TimeUnit;
+import javax.ws.rs.HttpMethod;
+
+/**
+ * Class to handle Api calls to System Service
+ */
+public class RemoteHttpClient {
+  private static final long TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(10);
+  private static final long RETRY_BASE_DELAY_MILLIS = 200L;
+  private static final long RETRY_MAX_DELAY_MILLIS = TimeUnit.SECONDS.toMillis(5);
+  private static final double RETRY_DELAY_MULTIPLIER = 1.2d;
+  private static final double RETRY_RANDOMIZE_FACTOR = 0.1d;
+
+  ServiceDiscoverer serviceDiscoverer;
+
+  public RemoteHttpClient(ServiceDiscoverer context) {
+    this.serviceDiscoverer = context;
+  }
+
+  public String retryableApiCall(String connectionUrl, String ...method) {
+    long delay = RETRY_BASE_DELAY_MILLIS;
+    double minMultiplier = RETRY_DELAY_MULTIPLIER - RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    double maxMultiplier = RETRY_DELAY_MULTIPLIER + RETRY_DELAY_MULTIPLIER * RETRY_RANDOMIZE_FACTOR;
+    Stopwatch stopWatch = new Stopwatch().start();
+
+    try {
+      while (stopWatch.elapsedTime(TimeUnit.MILLISECONDS) < TIMEOUT_MILLIS) {
+        try {
+          if (method[0].equals(HttpMethod.GET)) {
+            return executeGetCall(getHttpUrlConnection(connectionUrl));
+          } else if (method[0].equals(HttpMethod.POST) || method[0].equals(HttpMethod.PUT)) {
+            return executeCallWithBody(getHttpUrlConnection(connectionUrl), method[0], method[1]);
+          }
+        } catch (RetryableException | IOException e) {
+          TimeUnit.MILLISECONDS.sleep(delay);
+          delay = (long) (delay * (minMultiplier + Math.random() * (maxMultiplier - minMultiplier + 1)));
+          delay = Math.min(delay, RETRY_MAX_DELAY_MILLIS);
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Thread interrupted while trying to connect with State Store Service " +
+                                   "(AssessorService)", e);
+    }
+    throw new IllegalStateException("Timed out when trying to connect with State Store Service (AssessorService)");
+  }
+
+  private String executeGetCall(HttpURLConnection urlConn) throws IOException {
+    urlConn.setRequestMethod(HttpMethod.GET);
+    return validateAndRetrieveContent(urlConn);
+  }
+
+  private String executeCallWithBody(HttpURLConnection urlConn, String method, String body) throws IOException {
+
+    urlConn.setDoOutput(true);
+    urlConn.setRequestMethod(method);
+
+    OutputStream os = urlConn.getOutputStream();
+    OutputStreamWriter osw = new OutputStreamWriter(os, "UTF-8");
+    osw.write(body);
+    osw.flush();
+    osw.close();
```

Comment:
close() should usually be in a try-finally. An exception above will cause these not to be closed.